src/coreComponents/meshUtilities/InternalWellboreGenerator.cpp

```diff
@@ -0,0 +1,216 @@
+    setDescription( "Number of elements in the radial direction" );
+
+  registerWrapper( viewKeyStruct::tElemsString(), &m_tElems ).
+    setApplyDefaultValue( 40 ).
+    setSizedFromParent( 0 ).
+    setInputFlag( InputFlags::OPTIONAL ).
+    setDescription( "Number of elements in the tangent direction" );
+
+  registerWrapper( viewKeyStruct::rBiasString(), &m_rBias ).
+    setApplyDefaultValue( -0.8 ).
+    setSizedFromParent( 0 ).
+    setInputFlag( InputFlags::OPTIONAL ).
+    setDescription( "Bias of element sizes in the radial direction" );
+
+  registerWrapper( viewKeyStruct::trajectoryString(), &m_trajectory ).
+    setSizedFromParent( 0 ).
+    setInputFlag( InputFlags::OPTIONAL ).
+    setDescription( "Coordinates defining the wellbore trajectory" );
+}
+
+void InternalWellboreGenerator::postProcessInput()
+{
+  m_vertices[0].resize( 2 );
+  m_vertices[1].resize( 2 );
+  m_vertices[2].resize( 2 );
+  m_nElems[0].resize( 1 );
+  m_nElems[1].resize( 1 );
+  m_nElemBias[0].resize( 1 );
+
+  m_vertices[0][0]  = m_radius;
+  m_vertices[0][1]  = m_rOut;
+  m_vertices[1][0]  = 0;
+  m_vertices[1][1]  = m_theta;
+  m_vertices[2][0]  = m_trajectory[0][2];
+  m_vertices[2][1]  = m_trajectory[1][2];
+  m_nElems[0][0]    = m_rElems;
+  m_nElems[1][0]    = m_tElems;
+  m_nElemBias[0][0] = m_rBias;
+
+  InternalMeshGenerator::postProcessInput();
+}
+
+void InternalWellboreGenerator::generateMesh( DomainPartition & domain )
+{
+  InternalMeshGenerator::generateMesh( domain );
+
+  Group & meshBodies = domain.getGroup( string( "MeshBodies" ));
+  MeshBody & meshBody = meshBodies.registerGroup< MeshBody >( this->getName() );
+  MeshLevel & meshLevel0 = meshBody.registerGroup< MeshLevel >( string( "Level0" ));
+  NodeManager & nodeManager = meshLevel0.getNodeManager();
+  Group & nodeSets = nodeManager.sets();
+
+  // Wellbore nodesets
+  SortedArray< localIndex > & rnegNodes = nodeSets.registerWrapper< SortedArray< localIndex > >( string( "rneg" ) ).reference();
+  SortedArray< localIndex > & rposNodes = nodeSets.registerWrapper< SortedArray< localIndex > >( string( "rpos" ) ).reference();
+  SortedArray< localIndex > & tnegNodes = nodeSets.registerWrapper< SortedArray< localIndex > >( string( "tneg" ) ).reference();
+  SortedArray< localIndex > & tposNodes = nodeSets.registerWrapper< SortedArray< localIndex > >( string( "tpos" ) ).reference();
+
+  arrayView2d< real64, nodes::REFERENCE_POSITION_USD > const & X = nodeManager.referencePosition();
+  for( int localNodeIndex=0; localNodeIndex<nodeManager.size(); ++localNodeIndex )
+  {
+    // Get Cartesian coordinates of a reference centered vertical wellbore
+    real64 & xCoord = X( localNodeIndex, 0 );
+    real64 & yCoord = X( localNodeIndex, 1 );
+    real64 const & zCoord = X( localNodeIndex, 2 );
+
+    // Compute cylindrical coordinates of a reference centered vertical wellbore
+    real64 rCoord = sqrt( xCoord * xCoord + yCoord * yCoord );
+    real64 tCoord;
+
+    if( rCoord < 1e-10 )
```

Comment:
Please use the precision variable (make it `protected` if needed).

Reply:
done